### PR TITLE
Wireshark Dissector

### PR DESCRIPTION
**Pull request by Hornwitser**
A short article about a protocol dissector I've written for Wireshark. I figured I should ask the devs before publishing it so I'm waiting for a reply there. But either way you can have a look at it and say if it's a good fit for Alt-F4. It's rather technical in nature.  Feedback on any way to improve it is as always welcome.

### Author info
Name: Hornwitser
Suggested release date: Any time soon.

#### Requirements for media in submissions:

- [ ] Images are hosted on Media server 
  - Uploaded to server by alt-f4 editors, generally provided through discord or github

### CODE REVIEW

File: submissions/hornwitser_dissector/wireshark-dissector.md [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+In the past _Twinsen_ has posted about making a dissector for Wireshark in [FFF #196](https://factorio.com/blog/post/fff-196).
+The post contains quite a lot of internal information about the protocol which I could have used when I made my dissector, but I chose not to as I wanted to make mine without relying on internal information from the game source code.
+An obvious question right now though is why not use _Twinsen_'s dissector instead of writing my own?
+Well for a start it's not publicly available.
+But more importantly _it can't be made publicly available_.
+The way _Twinsen's_ dissector worked was to build Factorio as a shared library, and then build a custom version of Wireshark that linked to that shared library.
+But Wireshark is licensed under the GPL and you are not allowed to distribute GPL code that links with proprietary code, so to distribute the dissector Wube would have to release Factorio under the GPL which is obviously not going to happen.
+
+But a few weeks ago, I realized something.
+The Windows version of Factorio ships with a .pdb file.
```
Comment (heinwessels) [inline submissions/hornwitser_dissector/wireshark-dissector.md]: Why is Factorio shipped with a `.pdb` file? Even if it's a best guess. Could it be for the Factorio Mod Debugger?

**Comment by Hornwitser** [comment]
The text is now ready for review and editing.

File: submissions/hornwitser_dissector/wireshark-dissector.md [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+I could also get the enum definitions to translate numeric values to logical names for the fields that use them.
+And after some tidying up of the code and how the data was presented I was able to make the dissector output look like this:
+
+![](media/new-dissector.png)
+
+No longer does one have to guess what an "Odd Blue" value is, it's clearly it's the id of an InputActionSegment.
+And you can alse see the type is translated from the machine code looking 0xa2 into the much more friendly name "ServerCommand".
+But by doing this my dissector is no longer a cleanly separate thing from Factorio and how that affects the legallites around distributing this dissector I really have no clue.
+I asked the Factorio developers and they said they were fine with me releasing this to the public.
+So if you want to try it out yourself or add it to your toolbox you can find it [in my repository for it on GitHub](https://github.com/Hornwitser/factorio_dissector)
```
Comment (heinwessels) [inline submissions/hornwitser_dissector/wireshark-dissector.md]: In closing you could mention what you achieved with your tool. Did you find the issue why players were dropping, or did upi only figure it out too long afterwards? Or was it just a challenge in the end? And what can players achieve or learn by using this dissector?

File: submissions/hornwitser_dissector/wireshark-dissector.md [hunk 3]
```diff
@@ -0,0 +1,49 @@
+## Wireshark Dissector for Factorio <author>Hornwitser</author>
+
+This is a project that started around the Gridlock Cluster.
+I was investigating some issues with clients randomly dropping from the game and that brought me into examining packet traces using Wireshark.
```
Comment (heinwessels) [inline submissions/hornwitser_dissector/wireshark-dissector.md]: Maybe explain shortly what `packet traces` are.

**Review by heinwessels — COMMENTED** [review]
A really nice piece Hornwitser. Very interesting. I don't have many remarks, only where you can expand a little to help more readers understand.

File: submissions/hornwitser_dissector/wireshark-dissector.md (hunk 1 above)
Comment (Hornwitser) [reply submissions/hornwitser_dissector/wireshark-dissector.md]: It's used for symbolizing crash reports.

File: submissions/hornwitser_dissector/wireshark-dissector.md (hunk 3 above)
Comment (Hornwitser) [reply submissions/hornwitser_dissector/wireshark-dissector.md]: I expanded a bit on this section.

File: submissions/hornwitser_dissector/wireshark-dissector.md (hunk 2 above)
Comment (Hornwitser) [reply submissions/hornwitser_dissector/wireshark-dissector.md]: I tried to write something in the end for this. I'm not feeling like it fit well though.

Comment (heinwessels) [reply submissions/hornwitser_dissector/wireshark-dissector.md]: I think it came out good enough. If you're not happy with it you can always remove it again. Therenas will likely tweak it a little anyway.

**Comment by Nanogamer7** [comment]
**Fact Checking:** 

As you are presenting a program you made, the need for claims to be verified is limited, however I still listed the few details I found below

- ✔ Your description matches that of the [official website](https://www.wireshark.org/) and of the [Wikipedia page](https://en.wikipedia.org/wiki/Wireshark)
- ✔ Wireshark is licensed under GNU GPL 2 according to its [website](https://www.wireshark.org/docs/wsdg_html_chunked/AppGPL.html), and distributing it with propriatary Factorio code would violate section 3a) or 3b) (Disclaimer: I'm not a lawyer, nor in any other way qualified to give definitiv statements about legal documents etc.)
- ❔ Couldn't verify the `.pdb` file in the windows version, however it at least doesn't ship with the mac version (for obvious reasons)
- ✔ All links direct to where they say they are going, but maybe consider adding a link for the following:
  - [Alt-F4 #18 - Clustorio](https://alt-f4.blog/ALTF4-18/#the-road-to-clusterio-20-hornwitser) and also [Alt-F4 #18 - Modding](https://alt-f4.blog/ALTF4-18/#moddability-the-birth-of-a-mod-dedlyspyer) (how mods are loaded)
  - [cvdump.exe](https://github.com/Microsoft/microsoft-pdb/blob/master/cvdump/cvdump.exe)

---

Further notes:
- Some more formatting, especially `code` for filenames would improve readability

---

✔ Correct
❗ Incorrect
❔ Unclear. This can also mean I couldn't find any resources either proving or disproving a claim

Facts and Statement not listed above weren't validated by me; that doesn't mean they are false, it could be something personal (closed-source, etc.) I can't check, or you already provided a source. If you feel like a claim I left out should be validated, please tell me!

**Pull request merged**